### PR TITLE
add use_mkldnn attribute to ops in dygraph

### PR DESCRIPTION
**Pull request by sfraczek**
### PR types
New features

### PR changes
APIs

### Describe
Adds use_mkldnn argument to ops, similarly to how use_cudnn is used. In Dygraph mode, setting global FLAGS_use_mkldnn=true enables mkldnn for all supported operators. If it is desired to have non-mkldnn op when FLAGS_use_mkldnn=true, it can be done by passing use_mkldnn=False to Op as argument.

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -11322,7 +11322,7 @@ def scale(x, scale=1.0, bias=0.0, bias_after_scale=True, act=None, name=None):
     return helper.append_activation(out)
 
 
-def elementwise_add(x, y, axis=-1, act=None, name=None):
+def elementwise_add(x, y, axis=-1, act=None, name=None, use_mkldnn=False):
```
Comment (sfraczek) [inline python/paddle/fluid/layers/nn.py]: The fate of this line shall be decided after we resolve the issue discussed here: https://github.com/PaddlePaddle/Paddle/pull/25713#discussion_r461587882

File: python/paddle/fluid/dygraph/nn.py [hunk 2]
```diff
@@ -267,12 +283,14 @@ def forward(self, input):
...
-                attrs={'axis': 1})
+                attrs={'axis': 1,
+                       'use_mkldnn': self._use_mkldnn})
         else:
             pre_act = pre_bias
 
         # Currently, we don't support inplace in dygraph mode
-        return self._helper.append_activation(pre_act, act=self._act)
+        return self._helper.append_activation(
+            pre_act, act=self._act, use_mkl_dnn=self._use_mkldnn)
```
Comment (wojtuss) [inline python/paddle/fluid/dygraph/nn.py]: What about unifying the names and renaming `use_mkl_dnn` into `use_mkldnn`?

**Comment by grygielski** [comment]
@luotao1 As suggested here (https://github.com/PaddlePaddle/Paddle/pull/25713#issuecomment-668970545), I've changed a way oneDNN is called from `.../layers/nn.py/` to use `FLAGS_use_mkldnn`. Please let me know if you are fine with this solution.

**Comment by grygielski** [comment]
@luotao1 Could you send me logs of MAC_PR_CI_python35 or rerun if it's a random fail?

**Comment by luotao1** [comment]
```
[12:40:59][Step 1/1] 697/904 Test #700: test_slice_var .......................................................***Failed    2.98 sec
[12:40:59][Step 1/1] test_slice_var failed
[12:40:59][Step 1/1]  .E
[12:40:59][Step 1/1] ======================================================================
[12:40:59][Step 1/1] ERROR: test_check_output_8k (test_slice_var.TestSliceVar)
[12:40:59][Step 1/1] ----------------------------------------------------------------------
[12:40:59][Step 1/1] Traceback (most recent call last):
[12:40:59][Step 1/1]   File "/home/teamcity/work/37d0fbdf11d96284/build/python/paddle/fluid/tests/unittests/test_slice_var.py", line 61, in test_check_output_8k
[12:40:59][Step 1/1]     self.check_slice_output(shapes, expected_sizes, 8192)
[12:40:59][Step 1/1]   File "/home/teamcity/work/37d0fbdf11d96284/build/python/paddle/fluid/tests/unittests/test_slice_var.py", line 33, in check_slice_output
[12:40:59][Step 1/1]     shape=shape)
[12:40:59][Step 1/1]   File "/home/teamcity/work/37d0fbdf11d96284/build/python/paddle/fluid/framework.py", line 2640, in create_var
[12:40:59][Step 1/1]     var = Variable(block=self, *args, **kwargs)
[12:40:59][Step 1/1]   File "/home/teamcity/work/37d0fbdf11d96284/build/python/paddle/fluid/framework.py", line 930, in __init__
[12:40:59][Step 1/1]     "matched.".format(self.name, old_shape, shape))
[12:40:59][Step 1/1] ValueError: Variable 45858 has been created before. the previous shape is (8, 1020); the new shape is (800, 10). They are not matched.
[12:40:59][Step 1/1] 
[12:40:59][Step 1/1] ----------------------------------------------------------------------
[12:40:59][Step 1/1] Ran 2 tests in 0.003s
[12:40:59][Step 1/1] 
[12:40:59][Step 1/1] FAILED (errors=1)
```
It seems not random fail, but I rerun it again for you.

**Comment by lelelelelez** [comment]
@grygielski  you can download this log by https://paddle-docker-tar.cdn.bcebos.com/buildLog/386139.log

**Review by jczaja — DISMISSED** [review]
LGTM

**Comment by luotao1** [comment]
```
2020-08-06 20:00:32 
2020-08-06 20:00:34 ****************
2020-08-06 20:00:34 0. You must have one RD (XiaoguangHu01 or lanxianghit) and one TPM (saxon-zh or jzhang533 or swtkiwi or Heeenrrry or TCChenlong) approval for the api change for the management reason of API interface.
2020-08-06 20:00:34 1. You must have one TPM (saxon-zh or jzhang533 or swtkiwi or Heeenrrry or TCChenlong) approval for the api change for the management reason of API document.
2020-08-06 20:00:34 
2020-08-06 20:00:34 There are 2 approved errors.
2020-08-06 20:00:34 ****************
2020-08-06 20:00:34 API Difference is: 
2020-08-06 20:00:34 - paddle.fluid.dygraph.Conv2D (paddle.fluid.dygraph.nn.Conv2D, ('document', 'eaf7f009cbb4ce2f03453b767b8e3286'))
2020-08-06 20:00:34 ?                                                                            ^^   ^^ --    ^ ----- ^^^^ -----
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.fluid.dygraph.Conv2D (paddle.fluid.dygraph.nn.Conv2D, ('document', 'c1fc780ffd5769b4c16e71608c77fbf8'))
2020-08-06 20:00:34 ?                                                                            ^^ + ++ ^^^^^    ++ ^^^^^^^^  ^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.Conv2D.__init__ (ArgSpec(args=['self', 'num_channels', 'num_filters', 'filter_size', 'stride', 'padding', 'dilation', 'groups', 'param_attr', 'bias_attr', 'use_cudnn', 'act', 'dtype'], varargs=None, keywords=None, defaults=(1, 0, 1, None, None, None, True, None, 'float32')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.fluid.dygraph.Conv2D.__init__ (ArgSpec(args=['self', 'num_channels', 'num_filters', 'filter_size', 'stride', 'padding', 'dilation', 'groups', 'param_attr', 'bias_attr', 'use_cudnn', 'use_mkldnn', 'act', 'dtype'], varargs=None, keywords=None, defaults=(1, 0, 1, None, None, None, True, True, None, 'float32')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                                                                                               ++++++++++++++                                                                               ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.Pool2D (paddle.fluid.dygraph.nn.Pool2D, ('document', '1a697b10e08b443dc2865bfa1e1422a5'))
2020-08-06 20:00:34 ?                                                                               -----  ^^  ^ ----------------
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.fluid.dygraph.Pool2D (paddle.fluid.dygraph.nn.Pool2D, ('document', '28cc18bbe69ffb12e449a6e0344baebd'))
2020-08-06 20:00:34 ?                                                                            ++++ +++++++++++++++    ^  ^^^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.Pool2D.__init__ (ArgSpec(args=['self', 'pool_size', 'pool_type', 'pool_stride', 'pool_padding', 'global_pooling', 'use_cudnn', 'ceil_mode', 'exclusive', 'data_format'], varargs=None, keywords=None, defaults=(-1, 'max', 1, 0, False, True, False, True, 'NCHW')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.fluid.dygraph.Pool2D.__init__ (ArgSpec(args=['self', 'pool_size', 'pool_type', 'pool_stride', 'pool_padding', 'global_pooling', 'use_cudnn', 'use_mkldnn', 'ceil_mode', 'exclusive', 'data_format'], varargs=None, keywords=None, defaults=(-1, 'max', 1, 0, False, True, True, False, True, 'NCHW')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                                               ++++++++++++++                                                                                                            ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.Linear (paddle.fluid.dygraph.nn.Linear, ('document', 'db61dfab21ff2bd3984cccdfcb559570'))
2020-08-06 20:00:34 ?                                                                            ^^^^^^^^^^^^^^^ --   ^   ^^^ ^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.fluid.dygraph.Linear (paddle.fluid.dygraph.nn.Linear, ('document', '84844734ccb96fadf34fbce9f2a8acf0'))
2020-08-06 20:00:34 ?                                                                            ^^^^^^    ^^^^^  ++++ ^ ^^^^^^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.Linear.__init__ (ArgSpec(args=['self', 'input_dim', 'output_dim', 'param_attr', 'bias_attr', 'act', 'dtype'], varargs=None, keywords=None, defaults=(None, None, None, 'float32')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.fluid.dygraph.Linear.__init__ (ArgSpec(args=['self', 'input_dim', 'output_dim', 'param_attr', 'bias_attr', 'act', 'dtype', 'use_mkldnn'], varargs=None, keywords=None, defaults=(None, None, None, 'float32', True)), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                                ++++++++++++++                                                                      ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.BatchNorm (paddle.fluid.dygraph.nn.BatchNorm, ('document', 'e29cbd4c024975a79f1980cebacc4651'))
2020-08-06 20:00:34 ?                                                                                  ^  ^^^^^^^^^^^^^^    ^^ ---- ^^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.fluid.dygraph.BatchNorm (paddle.fluid.dygraph.nn.BatchNorm, ('document', '8882593f14f554039a93c664f7f8be43'))
2020-08-06 20:00:34 ?                                                                                  ^^^ + ^  +++++++ ++++++++++ ^  ^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.fluid.dygraph.BatchNorm.__init__ (ArgSpec(args=['self', 'num_channels', 'act', 'is_test', 'momentum', 'epsilon', 'param_attr', 'bias_attr', 'dtype', 'data_layout', 'in_place', 'moving_mean_name', 'moving_variance_name', 'do_model_average_for_mean_and_var', 'use_global_stats', 'trainable_statistics'], varargs=None, keywords=None, defaults=(None, False, 0.9, 1e-05, None, None, 'float32', 'NCHW', False, None, None, True, False, False)), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.fluid.dygraph.BatchNorm.__init__ (ArgSpec(args=['self', 'num_channels', 'act', 'is_test', 'momentum', 'epsilon', 'param_attr', 'bias_attr', 'dtype', 'data_layout', 'in_place', 'moving_mean_name', 'moving_variance_name', 'do_model_average_for_mean_and_var', 'use_global_stats', 'trainable_statistics', 'use_mkldnn'], varargs=None, keywords=None, defaults=(None, False, 0.9, 1e-05, None, None, 'float32', 'NCHW', False, None, None, True, False, False, True)), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                                                                                                                                                                                                                   ++++++++++++++                                                                                                                                      ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.nn.BatchNorm (paddle.fluid.dygraph.nn.BatchNorm, ('document', 'e29cbd4c024975a79f1980cebacc4651'))
2020-08-06 20:00:34 ?                                                                       ^  ^^^^^^^^^^^^^^    ^^ ---- ^^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.nn.BatchNorm (paddle.fluid.dygraph.nn.BatchNorm, ('document', '8882593f14f554039a93c664f7f8be43'))
2020-08-06 20:00:34 ?                                                                       ^^^ + ^  +++++++ ++++++++++ ^  ^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.nn.BatchNorm.__init__ (ArgSpec(args=['self', 'num_channels', 'act', 'is_test', 'momentum', 'epsilon', 'param_attr', 'bias_attr', 'dtype', 'data_layout', 'in_place', 'moving_mean_name', 'moving_variance_name', 'do_model_average_for_mean_and_var', 'use_global_stats', 'trainable_statistics'], varargs=None, keywords=None, defaults=(None, False, 0.9, 1e-05, None, None, 'float32', 'NCHW', False, None, None, True, False, False)), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.nn.BatchNorm.__init__ (ArgSpec(args=['self', 'num_channels', 'act', 'is_test', 'momentum', 'epsilon', 'param_attr', 'bias_attr', 'dtype', 'data_layout', 'in_place', 'moving_mean_name', 'moving_variance_name', 'do_model_average_for_mean_and_var', 'use_global_stats', 'trainable_statistics', 'use_mkldnn'], varargs=None, keywords=None, defaults=(None, False, 0.9, 1e-05, None, None, 'float32', 'NCHW', False, None, None, True, False, False, True)), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                                                                                                                                                                                                       ++++++++++++++                                                                                                                                       ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.nn.Pool2D (paddle.fluid.dygraph.nn.Pool2D, ('document', '1a697b10e08b443dc2865bfa1e1422a5'))
2020-08-06 20:00:34 ?                                                                    -----  ^^  ^ ----------------
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.nn.Pool2D (paddle.fluid.dygraph.nn.Pool2D, ('document', '28cc18bbe69ffb12e449a6e0344baebd'))
2020-08-06 20:00:34 ?                                                                 ++++ +++++++++++++++    ^  ^^^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.nn.Pool2D.__init__ (ArgSpec(args=['self', 'pool_size', 'pool_type', 'pool_stride', 'pool_padding', 'global_pooling', 'use_cudnn', 'ceil_mode', 'exclusive', 'data_format'], varargs=None, keywords=None, defaults=(-1, 'max', 1, 0, False, True, False, True, 'NCHW')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.nn.Pool2D.__init__ (ArgSpec(args=['self', 'pool_size', 'pool_type', 'pool_stride', 'pool_padding', 'global_pooling', 'use_cudnn', 'use_mkldnn', 'ceil_mode', 'exclusive', 'data_format'], varargs=None, keywords=None, defaults=(-1, 'max', 1, 0, False, True, True, False, True, 'NCHW')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                                    ++++++++++++++                                                                                                            ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.nn.Linear (paddle.fluid.dygraph.nn.Linear, ('document', 'db61dfab21ff2bd3984cccdfcb559570'))
2020-08-06 20:00:34 ?                                                                 ^^^^^^^^^^^^^^^ --   ^   ^^^ ^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 + paddle.nn.Linear (paddle.fluid.dygraph.nn.Linear, ('document', '84844734ccb96fadf34fbce9f2a8acf0'))
2020-08-06 20:00:34 ?                                                                 ^^^^^^    ^^^^^  ++++ ^ ^^^^^^^
2020-08-06 20:00:34 
2020-08-06 20:00:34 - paddle.nn.Linear.__init__ (ArgSpec(args=['self', 'input_dim', 'output_dim', 'param_attr', 'bias_attr', 'act', 'dtype'], varargs=None, keywords=None, defaults=(None, None, None, 'float32')), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 + paddle.nn.Linear.__init__ (ArgSpec(args=['self', 'input_dim', 'output_dim', 'param_attr', 'bias_attr', 'act', 'dtype', 'use_mkldnn'], varargs=None, keywords=None, defaults=(None, None, None, 'float32', True)), ('document', '6adf97f83acf6453d4a6a4b1070f3754'))
2020-08-06 20:00:34 ?                                                                                                                      ++++++++++++++                                                                     ++++++
2020-08-06 20:00:34 
2020-08-06 20:00:36 + approval_error=1
2020-08-06 20:00:36 + '[' 1 '!=' 0 ']'

```

File: python/paddle/fluid/dygraph/nn.py [hunk 3]
```diff
@@ -810,6 +831,7 @@ def __init__(self,
                  pool_padding=0,
                  global_pooling=False,
                  use_cudnn=True,
+                 use_mkldnn=True,
```
Comment (luotao1) [inline python/paddle/fluid/dygraph/nn.py]: Since we use `FLAGS_use_mkldnn` to global control, do you need to add `use_mkldnn` attribute in each API?

Comment (grygielski) [reply python/paddle/fluid/dygraph/nn.py]: As far as I understood our agreement, we didn't want to add use_mkldnn attribute in standard op API, thus we have decided to use Flags here. However for dygraph ops we wanted to be able to control every op whether it should be executed with oneDNN or not. Similarly these ops have use_cuda attribute. Am I wrong?

Comment (luotao1) [reply python/paddle/fluid/dygraph/nn.py]: @phlrain What's your opinion about it?

Comment (luotao1) [reply python/paddle/fluid/dygraph/nn.py]: > However for dygraph ops we wanted to be able to control every op whether it should be executed with oneDNN or not

@grygielski Discussed with @phlrain, we think `FLAGS_use_mkldnn` is enough for dygraph, we don't need to control every op whether it should be executed with oneDNN or not

**Comment by grygielski** [comment]
@luotao1 let me know if I should also delete `use_mkldnn` from `dygraph_utils.py` functions or it's fine like that.

**Comment by grygielski** [comment]
@luotao1 I've removed `use_mkldnn` attribute from all operators. I didn't delete it from `dygraph_utils.py`. Let me know if you are fine with that.

**Review by luotao1 — APPROVED** [review]
LGTM

**Review by wojtuss — APPROVED** [review]
LGTM

**Pull request merged**